File: src/kowalski/internal/note.py
import os

from datetime import datetime

from kowalski.internal.utils import colored, bold


TEMPLATE = """---
name: {}
created: {}
tags: {}
---

{}"""

class Note():

    def __init__(self, name: str, created: datetime, tags: list[str], content: str):

        self.name    = name
        self.created = created
        self.tags    = tags
        self.content = content

    def format(self):

        return TEMPLATE.format(self.name, self.created, self.tags, self.content)

    def write(self, path):

        f = open(os.path.join(path, self.name), "w")
        f.write(self.format())

    def __str__(self):

        name = self.name
        name = colored(235, 171, 52, name)
        name = bold(name)

        tags = str(self.tags)
        tags = tags.replace("[", "").replace("]", "")
        tags = tags.replace("'", "")
        tags = tags[:17]
        tags = tags.ljust(20, ".")
        tags = colored(0, 168, 138, tags)

        content = self.content
        content = content.replace("\n", " ")
        content = content[:27]
        content = content.ljust(30, ".")
        content = bold(content)

        return f"({name}): [🗞  {content}] [🏷️  {tags}]"

File: internal/model/note.go
package model

import (
	"strings"
	"fmt"
	"os"
	"path/filepath"
	"github.com/Zatfer17/kowalski-core/internal/utils"
)


var TEMPLATE = `---
created: %s
tags: [%s]
---
%s`

type Note struct {
	Created string
	Tags    []string
	Color   string
	Content string
}

func NewNote(created string, tags []string, content string) Note {
	return Note{
		Created: created,
		Tags:    tags,
		Color:   utils.GenerateColor(tags),
		Content: content,
	}
}

func (note Note) GetName() string {
	return fmt.Sprintf("%s.md", note.Created)
}

func (note Note) String() string {

	color := utils.ColoredSquare(note.Color)

	name := note.GetName()
	name  = utils.Colored(235, 171, 52, name)
	name  = utils.Bold(name)
	
	content := note.Content
	content  = strings.ReplaceAll(content, "\n", " ")
	if len(content) > 27 {
		content = content[:27]
	}
	content = content + strings.Repeat(".", 30-len(content))
	content = utils.Bold(content)

	tags := strings.Join(note.Tags, ",")
	if len(tags) > 17 {
		tags = tags[:17]
	}
	tags = tags + strings.Repeat(".", 20-len(tags))
	tags = utils.Colored(0, 168, 138, tags)

	return fmt.Sprintf("%s (%s): [🗞  %s] [🏷️  %s]", color, name, content, tags)
}

func (note Note) Format() string {
	return fmt.Sprintf(TEMPLATE, note.Created, strings.Join(note.Tags, ","), note.Content)
}

func (note Note) Write(directory string) error {

	filePath := filepath.Join(directory, note.GetName())

	f, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer f.Close()

	_, err = f.WriteString(note.Format())
	return err
}